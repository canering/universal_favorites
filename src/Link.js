import React, { useEffect } from "react";
import { useDrag, useDrop } from "react-dnd";
import { getEmptyImage } from "react-dnd-html5-backend";
import PropTypes from "prop-types";

function FolderSubDrop(props) {
    const [, drop] = useDrop({
        accept:"LINK",
        drop({ id: draggingId }){
            props.dropElement(props.id, draggingId, true);
        },
    });

    return(
        <span ref={drop}>
            <img src="https://img.icons8.com/color/48/000000/folder-invoices.png" alt="folder" className="icon" />
            {props.label}
        </span>
    );
}

FolderSubDrop.propTypes = {
    label:PropTypes.string,
    dropElement:PropTypes.func,
    id:PropTypes.number
};

function Link(props) {
    const [, drag, preview] = useDrag({
        item: { type: "LINK" },
        begin(){
            props.saveDraggingId(props.id);
            return {type: "LINK", label:props.item.label, link:props.item.link, draggingType:props.item.type, id:props.id};
        },
    });

    const [{isOver}, drop] = useDrop({
        accept:"LINK",
        drop({ id: draggingId }, monitor){
            if(!monitor.didDrop()) {
                props.dropElement(props.id, draggingId, false);
            }
        },
        collect: monitor => ({
            isOver: monitor.isOver({ shallow: true })
        }),
    });

    useEffect(() => {
        preview(getEmptyImage(), { captureDraggingState: true });
    });

    let style = null;
    if(isOver) {
        if(props.id < props.draggingId) {
            style = {borderTop:"2px dashed gray"};
        }
        else if(props.id > props.draggingId) {
            style = {borderBottom:"2px dashed gray"};
        }
    }

    let label = props.item.label;
    if(label.length > 20){
        label = label.substring(0, 19);
        label += "...";
    }

    if(props.item.type === "folder"){
        return (
            <div
                ref={node => drag(drop(node))}
                className="menu-item"
                style={style}
                onClick={() => props.openFolder(props.item.data, props.id)}
            >
                <FolderSubDrop label={props.item.label} dropElement={props.dropElement} id={props.id} />
                <img src="https://img.icons8.com/material/24/000000/sort-right--v1.png" alt="arrow" className="icon arrow" />
            </div>);
    }

    let url = props.item.link;
    if(!url.startsWith("http://") && !url.startsWith("https://")) {
        url = "http://" + url;
    }

    return(
        <div 
            ref={node => drag(drop(node))}
            className="menu-item" 
            style={style}
        >
            <a href={url} className="link" target="_blank" rel="noopener noreferrer">
                <img src={"https://www.google.com/s2/favicons?domain=" + props.item.link} alt="link" 
                    className="icon" 
                    onError={getFallbackFavicon} />
                {label}
            </a>
        </div>);
}

Link.propTypes = {
    item:PropTypes.object,
    openFolder:PropTypes.func,
    id:PropTypes.number,
    saveDraggingId:PropTypes.func,
    dropElement:PropTypes.func,
    draggingId:PropTypes.number
};

function getFallbackFavicon(event) {
    event.target.src = "https://img.icons8.com/material/24/000000/internet.png";
}

export default Link;