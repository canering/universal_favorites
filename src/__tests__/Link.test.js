import React from "react";
import {mount, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Link from "../Link";
import chrome from "sinon-chrome";
import { DndProvider } from "react-dnd";
import toJson from "enzyme-to-json";
import TestBackend from "react-dnd-test-backend";

window.chrome = chrome;

configure({adapter: new Adapter()});

const props = {
    openFolder: jest.fn(),
    id: 0,
    saveDraggingId: jest.fn(),
    dropElement: jest.fn(),
    draggingId: 0,
    setClicked: jest.fn(),
    setShiftClicked: jest.fn()
};

const linkItem = {
    type:"link",
    link:"http://example.com",
    label: "example.com"
};

const folderItem = {
    type:"folder",
    label: "example.com"
};

var linkComponent;
var folderComponent;

beforeEach(() => {
    linkComponent = mount(
        <DndProvider backend={TestBackend}>
            <Link item={linkItem} {...props} />
        </DndProvider>
    );

    folderComponent = mount(
        <DndProvider backend={TestBackend}>
            <Link item={folderItem} {...props} />
        </DndProvider>
    );
});

test("link snapshot test", () => {

    let tree = toJson(linkComponent);
    expect(tree).toMatchSnapshot();

    const clickedComponent = mount(
        <DndProvider backend={TestBackend}>
            <Link item={Object.assign({}, {clicked:true}, linkItem)} {...props} />
        </DndProvider>
    );

    tree = toJson(clickedComponent);
    expect(tree).toMatchSnapshot();
});

test("folder snapshot test", () => {
    let tree = toJson(folderComponent);
    expect(tree).toMatchSnapshot();

    const clickedComponent = mount(
        <DndProvider backend={TestBackend}>
            <Link item={Object.assign({}, {clicked:true}, folderItem)} {...props} />
        </DndProvider>
    );

    tree = toJson(clickedComponent);
    expect(tree).toMatchSnapshot();
});

test("when folder is clicked openFolder should be called", () => {

    folderComponent.find(".menu-item").prop("onClick")({});

    expect(props.openFolder).toHaveBeenCalledTimes(1);
    expect(props.openFolder).toHaveBeenCalledWith(folderItem, props.id);
});

test("when link is clicked current tab should be updated", () => {

    linkComponent.find(".menu-item").prop("onClick")({});

    expect(chrome.tabs.update.called).toBe(true);
});

test("when link or folder is ctrl clicked setClicked should be called", () => {

    linkComponent.find(".menu-item").prop("onClick")({ctrlKey:true});
    folderComponent.find(".menu-item").prop("onClick")({ctrlKey:true});

    expect(props.setClicked).toHaveBeenCalledTimes(2);
    expect(props.setClicked).toHaveBeenCalledWith(props.id);
});

test("when link or folder is ctrl clicked setShiftClicked should be called", () => {

    linkComponent.find(".menu-item").prop("onClick")({shiftKey:true});
    folderComponent.find(".menu-item").prop("onClick")({shiftKey:true});

    expect(props.setShiftClicked).toHaveBeenCalledTimes(2);
    expect(props.setShiftClicked).toHaveBeenCalledWith(props.id);
});