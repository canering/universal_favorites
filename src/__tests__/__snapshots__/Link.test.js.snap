// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`folder snapshot test 1`] = `
<DndProvider
  backend={[Function]}
>
  <Link
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "label": "example.com",
        "type": "folder",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      style={Object {}}
    >
      <FolderSubDrop
        dropElement={[MockFunction]}
        id={0}
        label="example.com"
      >
        <div
          className="menu-item-content"
        >
          <img
            alt="folder"
            className="icon"
            src="folder.png"
          />
          example.com
        </div>
      </FolderSubDrop>
      <img
        alt="arrow"
        className="icon arrow"
        src="right_arrow.png"
      />
    </div>
  </Link>
</DndProvider>
`;

exports[`folder snapshot test 2`] = `
<DndProvider
  backend={[Function]}
>
  <Link
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "clicked": true,
        "label": "example.com",
        "type": "folder",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "lightgray",
        }
      }
    >
      <FolderSubDrop
        dropElement={[MockFunction]}
        id={0}
        label="example.com"
      >
        <div
          className="menu-item-content"
        >
          <img
            alt="folder"
            className="icon"
            src="folder.png"
          />
          example.com
        </div>
      </FolderSubDrop>
      <img
        alt="arrow"
        className="icon arrow"
        src="right_arrow.png"
      />
    </div>
  </Link>
</DndProvider>
`;

exports[`link snapshot test 1`] = `
<DndProvider
  backend={[Function]}
>
  <Link
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "label": "example.com",
        "link": "http://example.com",
        "type": "link",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={Object {}}
    >
      <div
        className="menu-item-content"
      >
        <img
          alt="link"
          className="icon"
          hidden={false}
          src="default.png"
        />
        <img
          alt="link"
          className="icon"
          hidden={true}
          onLoad={[Function]}
          src="https://www.google.com/s2/favicons?domain=http://example.com"
        />
        example.com
      </div>
    </div>
  </Link>
</DndProvider>
`;

exports[`link snapshot test 2`] = `
<DndProvider
  backend={[Function]}
>
  <Link
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "clicked": true,
        "label": "example.com",
        "link": "http://example.com",
        "type": "link",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "backgroundColor": "lightgray",
        }
      }
    >
      <div
        className="menu-item-content"
      >
        <img
          alt="link"
          className="icon"
          hidden={false}
          src="default.png"
        />
        <img
          alt="link"
          className="icon"
          hidden={true}
          onLoad={[Function]}
          src="https://www.google.com/s2/favicons?domain=http://example.com"
        />
        example.com
      </div>
    </div>
  </Link>
</DndProvider>
`;
